### PR TITLE
bpo-47088: Add typing.LiteralString (PEP 675)

### PR DESCRIPTION
**Pull request by JelleZijlstra**
https://bugs.python.org/issue47088

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
cc @pradeep90 @gbleaney

File: Lib/typing.py [hunk 1]
```diff
@@ -510,6 +511,28 @@ def returns_self(self) -> Self:
     raise TypeError(f"{self} is not subscriptable")
 
 
+@_SpecialForm
+def LiteralString(self, parameters):
+    """Represents an arbitrary literal string.
+
+    Example::
+
+        from typing_extensions import LiteralString
```
Comment (ngnpope) [inline Lib/typing.py]:
```suggestion
        from typing import LiteralString
```

**Review by ngnpope — CHANGES_REQUESTED** [review]

File: Lib/typing.py (hunk 1 above)
Comment (JelleZijlstra) [reply Lib/typing.py]: Oops, that's embarrassing :) Thanks!

**Review by Fidget-Spinner — APPROVED** [review]

**Review by raghavthind2005 — APPROVED** [review]

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -578,6 +580,30 @@ These can be used as types in annotations and do not support ``[]``.
    * Every type is compatible with :data:`Any`.
    * :data:`Any` is compatible with every type.
 
+.. data:: LiteralString
+
+   Special type that includes only literal strings. A string
+   literal is compatible with ``LiteralString``, as is another
+   ``LiteralString``, but an object typed as just ``str`` is not.
```
Comment (pradeep90) [inline Doc/library/typing.rst:594]: @JelleZijlstra  Can we also point out that composing literal strings is fine too?

A one-line example below might be useful too.

Comment (gvanrossum) [reply Doc/library/typing.rst:594]: I didn't follow the PEP closely -- how far does this go? I suppose `"a" + "b"` is good. What about `"a" * 3`? Or `"a %s z" % "qqq"`? Or `",".join(("x", "y", "z"))`?

Comment (pradeep90) [reply Doc/library/typing.rst:594]: All three are fine. Basically, if all the inputs (including `self`) are `LiteralString`, then the output type is a `LiteralString`. (We [listed](https://peps.python.org/pep-0675/#appendix-c-str-methods-that-preserve-literalstring) `str` operations that preserve the `LiteralString` type.)

Comment (JelleZijlstra) [reply Doc/library/typing.rst:594]: Added

Comment (JelleZijlstra) [reply Doc/library/typing.rst:594]: @pradeep90 any other feedback? Thanks for the review!

File: Lib/typing.py [hunk 3]
```diff
@@ -523,6 +524,28 @@ def returns_self(self) -> Self:
...
+
+    Example::
+
+        from typing import LiteralString
+
+        def query(sql: LiteralString) -> ...:
+            ...
+
+        query("SELECT * FROM table")  # ok
+        query(f"SELECT * FROM {input()}")  # not ok
```
Comment (pradeep90) [inline Lib/typing.py]: It may be worth including the `+` example here too. Otherwise, LGTM!

**Review by pradeep90 — APPROVED** [review]

File: Lib/typing.py (hunk 3 above)
Comment (JelleZijlstra) [reply Lib/typing.py]: Thanks, done

**Pull request merged**